commit 8baf66cdb4e33d1b1738cdbf5619de6dcc1e8528 (refs/changes/62/762/1)
Author: Angela Fong <angela.fong@joyent.com>
Date:   2016-10-20T23:12:00-07:00 (3 years ago)
    
    ADMINUI-2346 Improved password expiration error handling
    ADMINUI-2347 Describe strong password guidelines on user profile page

diff --git a/lib/auth.js b/lib/auth.js
index 762e1cd2..6ba9c9ae 100644
--- a/lib/auth.js
+++ b/lib/auth.js
@@ -152,7 +152,10 @@ Auth.authenticate = function authenticate(req, res, next) {
         if (isLocked) {
             return next(new restify.ConflictError(ACCOUNT_LOCKED_MESSAGE));
         }
-
+        var pwdEndTime = user.pwdendtime;
+        if (pwdEndTime && pwdEndTime <= Date.now()) {
+            return next(new restify.ConflictError('Your password has already expired'));
+        }
         return doAuth();
     });
 
diff --git a/www/js/tpl/user-form.hbs b/www/js/tpl/user-form.hbs
index 4bd4cfcb..c8190f61 100644
--- a/www/js/tpl/user-form.hbs
+++ b/www/js/tpl/user-form.hbs
@@ -35,7 +35,9 @@
             <div class="form-group">
                 <label class="control-label col-xs-3">Password</label>
                 <div class="col-xs-7">
-                    <input type="password" placeholder="Change password" name="password" class="form-control" autocomplete="off">
+                    <input type="password" placeholder="Enter or update password" name="password" class="form-control" autocomplete="off">
+                    <a data-toggle="tooltip" data-placement="bottom" data-trigger="hover"
+                    title="Strong passwords should be at least 7-character long and contain numeric, uppercase, lowercase and non-alphabetic characters. Avoid the use of repeating or sequential characters. Do not re-use recent passwords.">Password guidelines</a>
                 </div>
             </div>
             <div class="form-group">
diff --git a/www/js/views/user-form.js b/www/js/views/user-form.js
index 36c834a5..63fa6c3d 100644
--- a/www/js/views/user-form.js
+++ b/www/js/views/user-form.js
@@ -154,6 +154,7 @@ module.exports = Backbone.Marionette.ItemView.extend({
         this.stickit();
         this.$('input:first').focus();
         this.$('.alert').hide();
+        this.$('[data-toggle="tooltip"]').tooltip();
         return this;
     }
 
